### ISSUE TITLE
Kotlin 2.0 and GradleInspector: Issues with variant selection

### ISSUE THREAD

**Issue by wkl3nk** [oss-review-toolkit/ort#8755]
Trying to scan an ordinary Kotlin 2.0 project, including Gradle-Kotlin and some standard Spring Boot 3.3.0 dependencies lead to error messages about _no matching variant of ... found_.

### To Reproduce

This ordinary project is here: https://github.com/wkl3nk/demo-7

### Console / log output

```
% ort -P ort.analyzer.enabledPackageManagers=GradleInspector analyze -i . -o .
 ______________________________
/        \_______   \__    ___/              The OSS Review Toolkit, version 22.0.1-001.sha.ce7d028,
|    |   | |       _/ |    |                 built with JDK 11.0.16.1+1, running under Java 21.0.3.
|    |   | |    |   \ |    |                 Executing 'analyze' as 'xxx' on Mac OS X
\________/ |____|___/ |____|                 with 12 CPUs and a maximum of 8192 MiB of memory.

Environment variables:
...
JAVA_HOME = /Users/xxx/Library/Java/JavaVirtualMachines/temurin-21.0.3/Contents/Home

Looking for ORT configuration in the following file:
        /Users/xxx/.ort/config/config.yml

Looking for analyzer-specific configuration in the following files and directories:
        /Users/xxx/osmi/demo-7/.ort.yml
        /Users/xxx/.ort/config/resolutions.yml (does not exist)
The following 1 package manager(s) are enabled:
        GradleInspector
The following 2 package curation provider(s) are enabled:
        DefaultDir, DefaultFile
Analyzing project path:
        /Users/xxx/osmi/demo-7
Found 1 GradleInspector definition file(s) at:
        build.gradle.kts
Found in total 1 definition file(s) from the following 1 package manager(s):
        GradleInspector
12:16:16.044 [DefaultDispatcher-worker-1] WARN  org.gradle.tooling.GradleConnector - Analyzing the project in '/Users/xxx/osmi/demo-7' produced the following error output:
	Could not resolve com.fasterxml.jackson.module:jackson-module-kotlin from project :. Causes are:
	org.gradle.internal.component.resolution.failure.exception.VariantSelectionException: No matching variant of com.fasterxml.jackson.module:jackson-module-kotlin:2.17.1 was found. The consumer was configured to find sources for use during runtime, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm' but:
	  - Variant 'apiElements' declares a component, packaged as a jar, and its dependencies declared externally:
	      - Incompatible because this component declares a library for use during compile-time and the consumer needed documentation for use during runtime
	      - Other compatible attributes:
	          - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
	          - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
	          - Doesn't say anything about the documentation type (required sources)
	  - Variant 'runtimeElements' declares a component for use during runtime, packaged as a jar, and its dependencies declared externally:
	      - Incompatible because this component declares a library and the consumer needed documentation
	      - Other compatible attributes:
	          - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
	          - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
	          - Doesn't say anything about the documentation type (required sources)
	Could not resolve com.fasterxml.jackson.module:jackson-module-kotlin from project :. Causes are:
	org.gradle.internal.component.resolution.failure.exception.VariantSelectionException: No matching variant of com.fasterxml.jackson.module:jackson-module-kotlin:2.17.1 was found. The consumer was configured to find sources for use during runtime, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally, as well as attribute 'documentation' with value 'documentation', attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm' but:
	  - Variant 'apiElements' declares a component, packaged as a jar, and its dependencies declared externally:
	      - Incompatible because this component declares a library for use during compile-time and the consumer needed documentation for use during runtime
	      - Other compatible attributes:
	          - Doesn't say anything about documentation (required 'documentation')
	          - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
	          - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
	          - Doesn't say anything about the documentation type (required sources)
	  - Variant 'runtimeElements' declares a component for use during runtime, packaged as a jar, and its dependencies declared externally:
	      - Incompatible because this component declares a library and the consumer needed documentation
	      - Other compatible attributes:
	          - Doesn't say anything about documentation (required 'documentation')
	          - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
	          - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
	          - Doesn't say anything about the documentation type (required sources)
	Could not resolve com.fasterxml.jackson.module:jackson-module-kotlin from project :. Causes are:
	org.gradle.internal.component.resolution.failure.exception.VariantSelectionException: No matching variant of com.fasterxml.jackson.module:jackson-module-kotlin:2.17.1 was found. The consumer was configured to find sources for use during runtime, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm' but:
	  - Variant 'apiElements' declares a component, packaged as a jar, and its dependencies declared externally:
	      - Incompatible because this component declares a library for use during compile-time and the consumer needed documentation for use during runtime
	      - Other compatible attributes:
	          - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
	          - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
	          - Doesn't say anything about the documentation type (required sources)
	  - Variant 'runtimeElements' declares a component for use during runtime, packaged as a jar, and its dependencies declared externally:
	      - Incompatible because this component declares a library and the consumer needed documentation
	      - Other compatible attributes:
	          - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
	          - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
	          - Doesn't say anything about the documentation type (required sources)
	Could not resolve com.fasterxml.jackson.module:jackson-module-kotlin from project :. Causes are:
	org.gradle.internal.component.resolution.failure.exception.VariantSelectionException: No matching variant of com.fasterxml.jackson.module:jackson-module-kotlin:2.17.1 was found. The consumer was configured to find sources for use during runtime, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally, as well as attribute 'documentation' with value 'documentation', attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm' but:
	  - Variant 'apiElements' declares a component, packaged as a jar, and its dependencies declared externally:
	      - Incompatible because this component declares a library for use during compile-time and the consumer needed documentation for use during runtime
	      - Other compatible attributes:
	          - Doesn't say anything about documentation (required 'documentation')
	          - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
	          - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
	          - Doesn't say anything about the documentation type (required sources)
	  - Variant 'runtimeElements' declares a component for use during runtime, packaged as a jar, and its dependencies declared externally:
	      - Incompatible because this component declares a library and the consumer needed documentation
	      - Other compatible attributes:
	          - Doesn't say anything about documentation (required 'documentation')
	          - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
	          - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
	          - Doesn't say anything about the documentation type (required sources)
...
```

### First ideas:
With Kotlin 2.0.0 as dependency in _build.gradle.kts_ - in comparison to Kotlin 1.9.24 - I see two additional configurations:
* dependencySources
* testDependencySources
 
Both of them carry an attribute _documentation_ that is mentioned in the error messages ("_consumer needed documentation_")

Scope exclusion of these two scopes in _.ort.yml_ did not solve the problem.

**Comment by wkl3nk** [comment]
Some more things to reproduce with the demo-7 project with configuration _dependencySources_:

```
% ./gradlew dependencyInsight --dependency jackson-module-kotlin --configuration dependencySources

> Task :dependencyInsight
com.fasterxml.jackson.module:jackson-module-kotlin: (selected by rule) FAILED
   Failures:
      - Could not resolve com.fasterxml.jackson.module:jackson-module-kotlin.
        No matching variant errors are explained in more detail at https://docs.gradle.org/8.8/userguide/variant_model.html#sub:variant-no-match.
          - No matching variant of com.fasterxml.jackson.module:jackson-module-kotlin:2.17.1 was found. The consumer was configured to find sources for use during runtime, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm' but:
              - Variant 'apiElements' declares a component, packaged as a jar, and its dependencies declared externally:
                  - Incompatible because this component declares a library for use during compile-time and the consumer needed documentation for use during runtime
                  - Other compatible attributes:
                      - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                      - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
                      - Doesn't say anything about the documentation type (required sources)
              - Variant 'runtimeElements' declares a component for use during runtime, packaged as a jar, and its dependencies declared externally:
                  - Incompatible because this component declares a library and the consumer needed documentation
                  - Other compatible attributes:
                      - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
                      - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'jvm')
                      - Doesn't say anything about the documentation type (required sources)

com.fasterxml.jackson.module:jackson-module-kotlin FAILED
\--- dependencySources
```
This fails with the well-known error message.
% ./gradlew dependencyInsight --dependency spring-boot-starter-web --configuration dependencySources
```
> Task :dependencyInsight
org.springframework.boot:spring-boot-starter-web:3.3.0 (selected by rule)
  Variant sourcesElements:
    | Attribute Name                     | Provided      | Requested     |
    |------------------------------------|---------------|---------------|
    | org.gradle.status                  | release       |               |
    | org.gradle.category                | documentation | documentation |
    | org.gradle.dependency.bundling     | external      | external      |
    | org.gradle.docstype                | sources       | sources       |
    | org.gradle.usage                   | java-runtime  | java-runtime  |
    | org.gradle.jvm.environment         |               | standard-jvm  |
    | org.gradle.libraryelements         |               | jar           |
    | org.jetbrains.kotlin.platform.type |               | jvm           |

org.springframework.boot:spring-boot-starter-web -> 3.3.0
\--- dependencySources
```

With dependency _spring-boot-starter-web_, everything looks all right.
So, on first sight, it seems there is something wrong with _jackson-kotlin-module_, because there are other dependencies on this configuration that work.

**Comment by wkl3nk** [comment]
As a workaround, these two new configurations _dependencySources_ and _testDependencySources_ could be just declared as non-relevant, as it was done already for another configuration where the name of the configuration ends with "DependenciesMetadata":

https://github.com/oss-review-toolkit/ort/blob/d42a87a0fcf528297e806b70b4d2f78492ab36c8/plugins/package-managers/gradle-plugin/src/main/kotlin/GradleModelExtensions.kt#L36-L54

But I am not sure that this is the way how to deal with this issue.

**Comment by sschuberth** [comment]
For the record, the solution mentioned at https://github.com/oss-review-toolkit/ort/issues/8377 does not seem to have a problem analyzing your example project, @wkl3nk.

**Comment by sschuberth** [comment]
> lead to error messages about _no matching variant of ... found_.

Actually, the log says it's just a *warning*:

```
12:16:16.044 [DefaultDispatcher-worker-1] WARN  org.gradle.tooling.GradleConnector
```

So does the result file actually get written correctly, just omitting `jackson-module-kotlin` from `dependencySources`? because if so, we'd be in line with Gradle's own plugin, see above.

**Comment by sschuberth** [comment]
> Actually, the log says it's just a _warning_:

Ok, we're a bit inconsistent here. Gradle writes the resolution issues to `stderr`, which ORT forwards to a `WARN` log:

https://github.com/oss-review-toolkit/ort/blob/f995050ff24808ffc95d77772963ca19d15f8141/plugins/package-managers/gradle-inspector/src/main/kotlin/GradleInspector.kt#L190-L195

Additionally, ORT maps error from Gradle buildig the dependency tree model to issues and logs errors:

https://github.com/oss-review-toolkit/ort/blob/f995050ff24808ffc95d77772963ca19d15f8141/plugins/package-managers/gradle-inspector/src/main/kotlin/GradleInspector.kt#L220-L222

I thing we should relax this a bit, and I'll probably propose a PR. Because I just verified that actually all works as expected (i.e. the `dependencySources` scope is present in the results will all dependencies that Gradle can resolve), and the error should simply be ignored.

**Comment by wkl3nk** [comment]
The customer project gets lots of _issues_, and it is **not just about this single dependency** to `jackson-module-kotlin`

I see this pattern many many times in the web app report:

```
When trying to find a matching variant:
      - Incompatible because this component declares a library for use during compile-time and the consumer needed documentation for use during runtime
      - Incompatible because this component declares a component of category 'verification' and the consumer needed documentation
      - Incompatible because this component declares a component of category 'dependency-analysis' and the consumer needed documentation
```

Yes, in fact it is just a warning.
I will exclude the Gradle configurations _dependencySources_ and _testDependencySources_ via scope excludes in _.ort.yml_. This will grey out these configurations in the web app report, nevertheless it still creates _issues_ in the report.
I also write a _resolution_ in _.ort.yml_ to ignore these kind of issues.

**Comment by sschuberth** [comment]
Meanwhile, I've created https://github.com/oss-review-toolkit/ort/pull/8781 and https://github.com/gradle/github-dependency-graph-gradle-plugin/issues/140.

**Comment by wkl3nk** [comment]
Meanwhile, we had a deeper look into one of our Kotlin 2.0 projects, and next to this ```VariantSelectionException```s we also have some  ```ModuleVersionNotFoundException```s due to the fact that for some reasons, when dependencies are declared at root level of the dependency tree - although they are transitive dependencies - and no version number is specified, these cannot be resolved, because the version number is missing. This may also be a follow-up error of ```VariantSelectionExceptions```in cases when Spring Boot "umbrella" dependencies like ```spring-boot-starter-test``` cannot be resolved. ```GradleInspector```then creates issues like 

_Could not find io.mockk:mockk:._
(Note the missing version number after the trailing colon)

This happens only on Gradle configuration that end in _dependencySources_, like

* _dependencySources_
* test_DependencySources_
* architectureTest_DependencySources_
* componentTest_DependencySources_
* generatedApiSources_DependencySources_
* integrationTest_DependencySources_
* localTestEnvironment_DependencySources_
* restDocs_DependencySources_
* testFixtures_DependencySources_
* ...

I am not expert enough to say if these configurations are relevant for OSS compliance, but maybe they could be declared as non-relevant in ```GradleInspector```.

**Comment by sschuberth** [comment]
I also did some further research, and [it indeed seems to be safe to ignore `dependencySources` configurations](https://github.com/JetBrains/kotlin/commit/73b290a94831a1aee547c99ab319fafc311864d7#diff-13384ab95f3c49314bc7e8d26b20c8981d5c514272bbd8c20130b9761be57fb6R724-R729), and [the issue will be fixed with Kotlin 2.0.20 eventually](https://github.com/JetBrains/kotlin/commit/659a6fbea22557c59f097acc94ec7f99f95e0bf6).

I'll prepare a PR for the `GradleInspector`.
